packages/hierarchies-react/src/presentation-hierarchies-react/UseSelectionHandler.ts

```diff
@@ -0,0 +1,135 @@
+  flatNodeList: Array<string>;
+  nodeIdToIndexMap: Map<string, number>;
+}
+
+/** @beta */
+export function useSelectionHandler(props: UseSelectionHandlerProps): UseSelectionHandlerResult {
+  const { rootNodes, selectionMode, selectNodes } = props;
+  const previousSelectionRef = useRef<string | undefined>(undefined);
+  const [state, setState] = useState<FlatTreeState>({
+    flatNodeList: [],
+    nodeIdToIndexMap: new Map(),
+  });
+
+  useEffect(() => {
+    if (!rootNodes) {
+      return;
+    }
+    setState(computeFlatNodeList(rootNodes));
+  }, [rootNodes]);
+
+  const getNodeRange = (firstId?: string, secondId?: string) => {
+    const getIndex = (nodeId?: string) => {
+      return nodeId ? state.nodeIdToIndexMap.get(nodeId) : 0;
+    };
+    const firstIndex = getIndex(firstId);
+    const secondIndex = getIndex(secondId);
+    if (firstIndex === undefined || secondIndex === undefined) {
+      return [];
+    }
+
+    const startingIndex = Math.min(firstIndex, secondIndex);
+    const endIndex = Math.max(firstIndex, secondIndex);
+    return state.flatNodeList.slice(startingIndex, endIndex + 1);
+  };
+
+  const onNodeSelect = (nodeId: string, isSelected: boolean, shiftDown: boolean, ctrlDown: boolean) => {
+    const selection = getSelection(selectionMode, isSelected, shiftDown, ctrlDown);
+    if (selection.type === "disabled") {
+      return;
+    }
+
+    const nodes = selection.select === "range" ? getNodeRange(previousSelectionRef.current, nodeId) : [nodeId];
+    selection.select !== "range" && (previousSelectionRef.current = nodeId);
+    selectNodes(nodes, selection.type);
+  };
+
+  const onNodeClick = (nodeId: string, isSelected: boolean, event: React.MouseEvent<HTMLElement, MouseEvent>) => {
+    return onNodeSelect(nodeId, isSelected, event.shiftKey, event.ctrlKey);
+  };
+
+  const onNodeKeyDown = (nodeId: string, isSelected: boolean, event: React.KeyboardEvent<HTMLElement>) => {
+    if (event.key === " " || event.key === "Spacebar" || event.key === "Enter") {
+      return onNodeSelect(nodeId, isSelected, event.shiftKey, event.ctrlKey);
+    }
+  };
+
+  return { onNodeClick, onNodeKeyDown };
+}
+
+function getSelection(
+  selectionMode: SelectionMode,
+  isSelected: boolean,
+  shiftDown: boolean,
+  ctrlDown: boolean,
+): { select: "node" | "range"; type: SelectionChangeType | "disabled" } {
+  if (selectionMode === "none") {
+    return { select: "node", type: "disabled" };
+  }
+
+  if (!isSelected && !(selectionMode === "extended" && shiftDown)) {
+    return { select: "node", type: "remove" };
+  }
+
+  switch (selectionMode) {
+    case "single":
+      return { select: "node", type: "replace" };
+    case "multiple":
+      return { select: "node", type: "add" };
+    case "extended":
+      if (shiftDown) {
+        return { select: "range", type: "replace" };
+      }
+      if (ctrlDown) {
+        return { select: "node", type: "add" };
+      }
+      return { select: "node", type: "replace" };
+  }
```

Comment:
Can you separate different selection modes handling to make it easier to read? I would suggest something like this:
switch(selectionMode) {
    case "node":
      return { select: "node", type: "disabled" };
    case "single":
      return { select: "node", type: "replace" };
    case "extended":
      return getExtendedSelectionAction();
    case "multiple":
      return getMultipleSelectionAction();
}

Reply:
That is definitely better, changed.